Shared/M365/GraphConnection.ps1

```diff
@@ -0,0 +1,178 @@
+
+.EXAMPLE
+$graphConnection = Connect-GraphAdvanced -Scopes Group.Read.All, User.Read.All -Modules Microsoft.Graph.Users, Microsoft.Graph.Groups
+This example establishes a connection to Microsoft Graph with the scopes "Group.Read.All" and "User.Read.All" using the "Microsoft.Graph.Users" and "Microsoft.Graph.Groups" modules.
+
+.EXAMPLE
+$graphConnection = Connect-GraphAdvanced -Scopes Group.Read.All, User.Read.All -Modules Microsoft.Graph.Users, Microsoft.Graph.Groups -minModuleVersion 2.25.0
+This example establishes a connection to Microsoft Graph with the scopes "Group.Read.All" and "User.Read.All" using the "Microsoft.Graph.Users" and "Microsoft.Graph.Groups" modules, and specifies a minimum module version of 2.25.0.
+#>
+
+. $PSScriptRoot\..\ModuleHandle.ps1
+
+function Connect-GraphAdvanced {
+    param (
+        [Parameter(Mandatory = $true)]
+        [string[]]$Scopes,
+        [Parameter(Mandatory = $true)]
+        [string[]]$Modules,
+        [Parameter(Mandatory = $false)]
+        [string[]]$TenantId = $null,
+        [Parameter(Mandatory = $false)]
+        [switch]$DoNotShowConnectionDetails,
+        [Parameter(Mandatory = $false)]
+        [System.Version]$MinModuleVersion = 2.0.0.0
+    )
+
+    #Validate Graph is installed and loaded
+    $requestModule = Request-Module -Modules $Modules -MinModuleVersion $MinModuleVersion
+    if (-not $requestModule) {
+        Write-Host "We cannot continue without $Modules Powershell module" -ForegroundColor Red
+        return $null
+    }
+
+    #Validate Graph is connected or try to connect
+    $connection = $null
+    try {
+        $connection = Get-MgContext -ErrorAction Stop
+    } catch {
+        Write-Host "We cannot check context. Error:`n$_" -ForegroundColor Red
+        return $null
+    }
+
+    if ($null -eq $connection) {
+        Write-Host "Not connected to Graph" -ForegroundColor Yellow
+        $connection = Add-GraphConnection -Scopes $Scopes
+    } else {
+        Write-Verbose "You have a Graph sessions"
+        Write-Verbose "Checking scopes"
+        if (-not (Test-GraphScopeContext -Scopes $connection.Scopes -ExpectedScopes $Scopes)) {
+            Write-Host "Not connected to Graph with expected scopes" -ForegroundColor Yellow
+            $connection = Add-GraphConnection -Scopes $Scopes
+        } else {
+            Write-Verbose "All scopes are present"
+        }
+    }
+
+    if ($connection) {
+        Write-Verbose "Checking TenantId"
+        if ($TenantId) {
+            if ($connection.TenantId -ne $TenantId) {
+                Write-Host "Connected to $($connection.TenantId). Not expected tenant: $TenantId" -ForegroundColor Red
+                return $null
+            } else {
+                Write-Verbose "TenantId is correct"
+            }
+        }
+
+        if (-not $DoNotShowConnectionDetails) {
+            $connection.PSObject.Properties | ForEach-Object { Write-Verbose "$($_.Name): $($_.Value)" }
```

Comment:
Move this to outside the `if` statement. We should always display this type of information in verbose.